src/core/thread/dua_manager.cpp

```diff
@@ -184,6 +249,522 @@ otError DuaManager::Store(void)
+
+        if (mDelay.mFields.mReregistrationDelay == 0)
+        {
+#if OPENTHREAD_CONFIG_DUA_ENABLE
+            if (mDuaState != kNotExist)
+            {
+                mDuaState = kToRegister;
+            }
+#endif
+
+#if OPENTHREAD_CONFIG_TMF_PROXY_DUA_ENABLE
+            mChildDuaRegisterMask.Clear();
+#endif
+            attempt = true;
+        }
+    }
+
+    if (attempt)
+    {
+        PerformNextRegistration();
+    }
+
+    ScheduleTimer();
+}
+
+void DuaManager::ScheduleTimer(void)
+{
+    if (mDelay.mValue == 0)
+    {
+        mTimer.Stop();
+    }
+    else if (!mTimer.IsRunning())
+    {
+        mTimer.Start(kStateUpdatePeriod);
+    }
+}
+
+void DuaManager::HandleRegistrationTask(Tasklet &aTasklet)
+{
+    aTasklet.GetOwner<DuaManager>().PerformNextRegistration();
+}
+
+void DuaManager::PerformNextRegistration(void)
+{
+    otError          error   = OT_ERROR_NONE;
+    Mle::MleRouter & mle     = Get<Mle::MleRouter>();
+    Coap::Message *  message = nullptr;
+    Ip6::MessageInfo messageInfo;
+    Ip6::Address     dua;
+
+    VerifyOrExit(mle.IsAttached(), OT_NOOP);
+    VerifyOrExit(Get<BackboneRouter::Leader>().HasPrimary(), OT_NOOP);
+
+    // Only allow one outgoing DUA.req
+    VerifyOrExit(!mIsDuaPending, OT_NOOP);
+
+    // Only send DUA.req when necessary
+#if OPENTHREAD_CONFIG_DUA_ENABLE
+#if OPENTHREAD_CONFIG_TMF_PROXY_DUA_ENABLE
+    VerifyOrExit(mle.IsRouterOrLeader() || !mle.IsExpectedToBecomeRouter(), OT_NOOP);
+    VerifyOrExit((mDuaState == kToRegister && mDelay.mFields.mRegistrationDelay == 0) ||
+                     mChildDuaMask != mChildDuaRegisterMask,
+                 OT_NOOP);
+#else
+    VerifyOrExit(mDuaState == kToRegister && mDelay.mFields.mRegistrationDelay == 0, OT_NOOP);
+#endif // OPENTHREAD_CONFIG_TMF_PROXY_DUA_ENABLE
+
+    VerifyOrExit(mle.IsFullThreadDevice() || mle.GetParent().IsThreadVersion1p1(), OT_NOOP);
+#endif // OPENTHREAD_CONFIG_DUA_ENABLE
+
+    // Prepare DUA.req
+    VerifyOrExit((message = Get<Coap::Coap>().NewPriorityMessage()) != nullptr, error = OT_ERROR_NO_BUFS);
+
+    SuccessOrExit(error =
+                      message->Init(OT_COAP_TYPE_CONFIRMABLE, OT_COAP_CODE_POST, OT_URI_PATH_DUA_REGISTRATION_REQUEST));
+    SuccessOrExit(error = message->SetPayloadMarker());
+
+#if OPENTHREAD_CONFIG_DUA_ENABLE
+    if (mDuaState == kToRegister && mDelay.mFields.mRegistrationDelay == 0)
+    {
+        dua = GetDomainUnicastAddress();
+        SuccessOrExit(error = Tlv::AppendTlv(*message, ThreadTlv::kTarget, &dua, sizeof(dua)));
+        SuccessOrExit(error = Tlv::AppendTlv(*message, ThreadTlv::kMeshLocalEid, &mle.GetMeshLocal64().GetIid(),
+                                             sizeof(Ip6::InterfaceIdentifier)));
+        mDuaState             = kRegistering;
+        mLastRegistrationTime = TimerMilli::GetNow();
+    }
+    else
+#endif // OPENTHREAD_CONFIG_DUA_ENABLE
+    {
+#if OPENTHREAD_CONFIG_TMF_PROXY_DUA_ENABLE
+        if (!mRegisterCurrentChildIndex)
```

Comment:
I have doubt about this code block. 
Seems the parent is choosing a Child that has Dua not registered to send DUA.req. But it seems to always find the first/last Child. If this Child failed in registering, how will the parent proceed to other Children?